sources/api/migration/migrator/src/main.rs

```diff
@@ -309,29 +308,36 @@ where
             output.status.success(),
             error::MigrationFailureSnafu { output }
         );
+
+        // If an intermediate datastore exists from a previous loop, delete it.
+        delete_intermediate_datastore(&intermediate_datastore);
+        // Remember the location of the target_datastore to delete it in the next loop iteration
+        // (i.e if it was an intermediate).
+        intermediate_datastore = Some(target_datastore.clone());
         source_datastore = &target_datastore;
     }
 
-    // Remove the intermediate data stores
-    intermediate_datastores.remove(&target_datastore);
-    for intermediate_datastore in intermediate_datastores {
-        // Even if we fail to remove an intermediate data store, we've still migrated
-        // successfully, and we don't want to fail the upgrade - just let someone know for
-        // later cleanup.
-        trace!(
-            "Removing intermediate data store at {}",
-            intermediate_datastore.display()
+    Ok(target_datastore)
+}
+
+// Try to delete an intermediate datastore if it exists. If it fails to delete, print an error.
+fn delete_intermediate_datastore(path: &Option<PathBuf>) {
+    let path = if let Some(p) = path {
+        p
+    } else {
+        // nothing to do
+        return;
+    };
```

Comment:
nit: I prefer moving the option checking outside of the scope of the `delete` function so the function does what it's literally named for and pares the diff a bit.
```suggestion
// Try to delete an intermediate datastore. If it fails to delete, print an error.
fn delete_intermediate_datastore(path: &PathBuf) {
```

Then on line 312:
```
// If an intermediate datastore exists from a previous loop, delete it.
if let Some(path) = intermediate_datastore {
  delete_intermediate_datastore(&path);
}
```